src/components/frontend-engine/frontend-engine.tsx

```diff
@@ -31,32 +32,40 @@ const FrontendEngineInner = forwardRef<IFrontendEngineRef, IFrontendEngineProps>
 		defaultValues: defaultValues,
 		resolver: yupResolver(validationSchema),
 	});
-	const { watch, handleSubmit: reactFormHookSubmit, formState } = formMethods;
-
-	// TODO: Remove logging
-	useEffect(() => {
-		const subscription = watch((value, { name, type }) => console.log(value, name, type));
-
-		return () => subscription.unsubscribe();
-	}, [watch]);
+	const { watch, handleSubmit: reactFormHookSubmit, getValues } = formMethods;
 
 	// =============================================================================
 	// HELPER FUNCTIONS
 	// =============================================================================
 	useImperativeHandle<Partial<IFrontendEngineRef>, Partial<IFrontendEngineRef>>(ref, () => ({
-		getFormState: () => {
-			return formState;
-		},
-		submit: () => {
-			reactFormHookSubmit(handleSubmit)();
-		},
+		getValues,
+		isValid: checkIsFormValid,
+		submit: reactFormHookSubmit(handleSubmit),
 	}));
 
+	const checkIsFormValid = useCallback(() => {
+		try {
+			validationSchema.validateSync(getValues());
+			return true;
+		} catch (error) {
+			return false;
+		}
+	}, [getValues, validationSchema]);
+
 	const handleSubmit = (data: TFrontendEngineValues) => {
 		console.log(data); // TODO: remove
 		onSubmit?.(data);
 	};
 
+	// =============================================================================
+	// EFFECTS
+	// =============================================================================
+	useEffect(() => {
+		const subscription = watch((value) => onChange?.(value, checkIsFormValid()));
```

Comment:
maybe we check if `onChange` specified instead of using the `?` denotation

Reply:
just curious, what are the cons of optional chaining a function?

Reply:
no obvious cons other than not even running the function if `onChange` is not specified.